MapboxVision/src/main/java/com/mapbox/vision/utils/observable/CompositeListener.kt

```diff
@@ -0,0 +1,45 @@
+package com.mapbox.vision.utils.observable
+
+import com.mapbox.vision.mobile.core.utils.delegate.DelegateWeakRef
+import java.lang.ref.WeakReference
+import java.util.concurrent.CopyOnWriteArraySet
+
+interface CompositeListener<T> {
+    fun addListener(observer: T)
+    fun removeListener(observer: T)
+
+    open class Impl<T> : CompositeListener<T> {
+        private val listeners = CopyOnWriteArraySet<WeakReference<T>>()
+        private var needClean = false
+
+        override fun addListener(observer: T) {
+            listeners.add(WeakReference(observer))
+        }
+
+        override fun removeListener(observer: T) {
+            listeners.find { it.get() == observer }?.let { listeners.remove(it) }
+        }
+
+        protected fun forEach(func: T.() -> Unit) {
+            listeners.forEach { weak ->
+                val instance = weak.get()
+                if (instance != null) {
+                    instance.func()
+                } else {
+                    listeners.remove(weak)
+                }
+            }
+        }
+    }
+}
+
+fun <T> delegateWeakPropertyObservable(
```

Comment:
Rename to smth like `weakCompositeListeners`?

Reply:
it's just listener, not composite. better `delegateWeakPropertyListener`

Reply:
Ok, just not `Observable`